### PR TITLE
PIC-2046 add endpoint replay messages from dlq

### PR DESCRIPTION
**Pull request by rschallapalli**

### CODE REVIEW

File: src/main/java/uk/gov/justice/probation/courtcasematcher/controller/SqsAdminController.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+public class SqsAdminController {
+
+    private SqsAdminService sqsAdminService;
+
+    public SqsAdminController(SqsAdminService sqsAdminService) {
+        this.sqsAdminService = sqsAdminService;
+    }
+
+    @PutMapping("/retry-all-dlqs")
+    public void replayDlqMessages() {
```
Comment (jhevans) [inline src/main/java/uk/gov/justice/probation/courtcasematcher/controller/SqsAdminController.java:17]: Will need auth on this if it isn't configured elsewhere. Example [here](https://github.com/ministryofjustice/hmpps-spring-boot-sqs/blob/2fa3ca29f8784d99b5e1d798a47557385c05def8/hmpps-sqs-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueueResource.kt#L16)

File: src/main/java/uk/gov/justice/probation/courtcasematcher/service/SqsAdminService.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        var messageCount = ofNullable(dlqAttributes.getAttributes())
+                .map(attrMap -> attrMap.get(APPROXIMATE_NUMBER_OF_MESSAGES_ATTR_NAME))
+                .map(Integer::parseInt)
+                .orElse(0);
+
+        log.info("Found {} messages on the dlq", messageCount);
+
+        var replayedMessages = 0;
+
+        while (replayedMessages < messageCount) {
```
Comment (jhevans) [inline src/main/java/uk/gov/justice/probation/courtcasematcher/service/SqsAdminService.java:57]: You could get rid of the while loop and use a `IntStream.range(0, messageCount).forEach()`  here 🤷  Personal preference really, I generally prefer streams but I'm fine with either here.

File: src/test/java/uk/gov/justice/probation/courtcasematcher/controller/SqsAdminControllerIntTest.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    @Test
+    void givenThereAreMessagesOnDlq_whenRetryAllDlqInvoked_shouldReplayMessages() {
+
+        purgeQueues();
+
+        sendMessageToDlq("message body 1");
+        sendMessageToDlq("message body 2");
+
+        var response = testRestTemplate.exchange("http://localhost:" + port + "/retry-all-dlqs", HttpMethod.PUT, null, String.class);
```
Comment (jhevans) [inline src/test/java/uk/gov/justice/probation/courtcasematcher/controller/SqsAdminControllerIntTest.java:61]: We generally use RestAssured rather than RestTemplate, there's an example [here](https://github.com/ministryofjustice/court-case-matcher/blob/9402aa735626ef579676e4c86c5e63d89df92b8b/src/test/java/uk/gov/justice/probation/courtcasematcher/health/HealthCheckTest.java#L51). Tends to read a bit better IMO but we should probably stick with it for consistency.

File: src/test/java/uk/gov/justice/probation/courtcasematcher/health/HealthCheckTest.java [hunk 4]
```diff
@@ -33,6 +35,9 @@
     @LocalServerPort
     private int port;
 
+    @MockBean
+    private SqsAdminService sqsAdminService;
```
Comment (jhevans) [inline src/test/java/uk/gov/justice/probation/courtcasematcher/health/HealthCheckTest.java:39]: Why is this needed?

File: src/test/java/uk/gov/justice/probation/courtcasematcher/messaging/SqsMessageReceiverIntTest.java [hunk 5]
```diff
@@ -228,17 +232,19 @@ public void givenExistingCase_whenReceivePayloadForOrganisation_thenSendUpdatedC
         verifyNoMoreInteractions(telemetryService);
     }
 
+
+
     @TestConfiguration
     public static class AwsTestConfig {
-        @Value("${aws.sqs.court_case_matcher_endpoint_url}")
+        @Value("${aws_sqs_court_case_matcher_endpoint_url}")
```
Comment (jhevans) [inline src/test/java/uk/gov/justice/probation/courtcasematcher/messaging/SqsMessageReceiverIntTest.java:239]: Could use relaxed binding again here

File: src/test/java/uk/gov/justice/probation/courtcasematcher/messaging/SqsMessageReceiverIntTest.java [hunk 6]
```diff
@@ -228,17 +232,19 @@ public void givenExistingCase_whenReceivePayloadForOrganisation_thenSendUpdatedC
...
         private String sqsEndpointUrl;
         @Value("${aws.access_key_id}")
         private String accessKeyId;
         @Value("${aws.secret_access_key}")
         private String secretAccessKey;
         @Value("${aws.region_name}")
         private String regionName;
-        @Value("${aws.sqs.court_case_matcher_queue_name}")
+        @Value("${aws_sqs_court_case_matcher_queue_name}")
```
Comment (jhevans) [inline src/test/java/uk/gov/justice/probation/courtcasematcher/messaging/SqsMessageReceiverIntTest.java:247]: same

File: src/test/resources/application-sqsadmin.yml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+aws_sqs_court_case_matcher_access_key_id: id1
```
Comment (jhevans) [inline src/test/resources/application-sqsadmin.yml:1]: Relaxed binding

File: src/test/resources/application-test.yml [hunk 8]
```diff
@@ -37,12 +37,15 @@ aws:
   access_key_id: foobar
   secret_access_key: foobar
   sqs:
-    court_case_matcher_endpoint_url: "http://localhost:4566"
```
Comment (jhevans) [inline src/test/resources/application-test.yml:40]: This change shouldn't have been necessary again because of relaxed binding!

**Review by jhevans — CHANGES_REQUESTED** [review]
All the important stuff looks good but I have left a few comments for review. A couple of optionals around code style and  a lot of changes to @Values injections which I don't think should be necessary due to [relaxed binding](https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.external-config.typesafe-configuration-properties.relaxed-binding).

File: src/main/java/uk/gov/justice/probation/courtcasematcher/messaging/SqsMessageReceiver.java [hunk 9]
```diff
@@ -33,14 +33,14 @@
     @NonNull
     private final TelemetryService telemetryService;
 
-    @Value("${aws.sqs.court_case_matcher_queue_name}")
+    @Value("${aws_sqs_court_case_matcher_queue_name}")
```
[the remark this answers is not in the document]
Comment (rschallapalli) [reply src/main/java/uk/gov/justice/probation/courtcasematcher/messaging/SqsMessageReceiver.java]: Done

File: src/main/java/uk/gov/justice/probation/courtcasematcher/messaging/SqsMessageReceiver.java [hunk 10]
```diff
@@ -33,14 +33,14 @@
...
     private String queueName;
 
     @Autowired
     @NonNull
     private CourtCaseExtractor courtCaseExtractor;
 
-    @SqsListener(value = "${aws.sqs.court_case_matcher_queue_name}", deletionPolicy = SqsMessageDeletionPolicy.ON_SUCCESS)
+    @SqsListener(value = "${aws_sqs_court_case_matcher_queue_name}", deletionPolicy = SqsMessageDeletionPolicy.ON_SUCCESS)
```
[the remark this answers is not in the document]
Comment (rschallapalli) [reply src/main/java/uk/gov/justice/probation/courtcasematcher/messaging/SqsMessageReceiver.java]: done

File: src/main/java/uk/gov/justice/probation/courtcasematcher/service/SqsAdminService.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    @Qualifier("courtCaseMatcherSqsDlq")
+    private AmazonSQSAsync courtCaseMatcherSqsQueue;
+
+
+    private String courtCaseMatcherSqsDlqUrl;
+
+    private String courtCaseMatcherSqsUrl;
+
+    public SqsAdminService(AmazonSQSAsync courtCaseMatcherSqsDlq, AmazonSQSAsync courtCaseMatcherSqsQueue,
+                           @Value("${aws_sqs_court_case_matcher_dlq_endpoint_url}") String courtCaseMatcherSqsDlqUrl,
```
[the remark this answers is not in the document]
Comment (rschallapalli) [reply src/main/java/uk/gov/justice/probation/courtcasematcher/service/SqsAdminService.java]: done

File: src/main/java/uk/gov/justice/probation/courtcasematcher/controller/SqsAdminController.java (hunk 1 above)
Comment (jhevans) [reply src/main/java/uk/gov/justice/probation/courtcasematcher/controller/SqsAdminController.java:20]: Raised [PIC-2052](https://dsdmoj.atlassian.net/browse/PIC-2052) to follow up with this

File: src/test/java/uk/gov/justice/probation/courtcasematcher/controller/SqsAdminControllerIntTest.java [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+    @Test
+    void givenThereAreMessagesOnDlq_whenRetryAllDlqInvoked_shouldReplayMessages() {
+
+        purgeQueues();
+
+        sendMessageToDlq("message body 1");
+        sendMessageToDlq("message body 2");
+
+        String sqsAdminUrl = "http://localhost:" + port + "/queue-admin/retry-all-dlqs";
```
Comment (jhevans) [inline src/test/java/uk/gov/justice/probation/courtcasematcher/controller/SqsAdminControllerIntTest.java:61]: Would generally prefer String.format()

File: src/main/java/uk/gov/justice/probation/courtcasematcher/service/SqsAdminService.java [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+@Service
+@Slf4j
+public class SqsAdminService {
+    private static final String APPROXIMATE_NUMBER_OF_MESSAGES_ATTR_NAME = "ApproximateNumberOfMessages";
+
+    @Autowired
+    @Qualifier("courtCaseMatcherSqsQueue")
+    private AmazonSQSAsync courtCaseMatcherSqsDlq;
+
+    @Autowired
```
[the remark this answers is not in the document]
Comment (rschallapalli) [reply src/main/java/uk/gov/justice/probation/courtcasematcher/service/SqsAdminService.java]: done

File: src/test/java/uk/gov/justice/probation/courtcasematcher/health/HealthCheckTest.java (hunk 4 above)
Comment (rschallapalli) [reply src/test/java/uk/gov/justice/probation/courtcasematcher/health/HealthCheckTest.java:39]: This test runs with "test" profile and SqsMessagingConfig is configured to load when profile is not "test" so SqsAdminService cannot be constructed as it has beans coming from SqsMessagingConfig.

File: src/test/java/uk/gov/justice/probation/courtcasematcher/controller/SqsAdminControllerIntTest.java [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        var response = testRestTemplate.exchange(sqsAdminUrl, HttpMethod.PUT, null, String.class);
```
Comment (jhevans) [inline src/test/java/uk/gov/justice/probation/courtcasematcher/controller/SqsAdminControllerIntTest.java:62]: Could we use rest assured here for consistency and readability. Something like:

```
given()
                .contentType(ContentType.JSON)
                .accept(ContentType.JSON)
                .when()
                .put(sqsAdminUrl)
                .then()
                .statusCode(201)
```

**Review by jhevans — CHANGES_REQUESTED** [review]
One usage of restTemplate which can be replaced

File: src/test/java/uk/gov/justice/probation/courtcasematcher/health/HealthCheckTest.java (hunk 4 above)
Comment (rschallapalli) [reply src/test/java/uk/gov/justice/probation/courtcasematcher/health/HealthCheckTest.java]: Moved to common config

File: src/test/java/uk/gov/justice/probation/courtcasematcher/messaging/SqsMessageReceiverIntTest.java (hunk 6 above)
Comment (rschallapalli) [reply src/test/java/uk/gov/justice/probation/courtcasematcher/messaging/SqsMessageReceiverIntTest.java]: addressed

File: src/test/resources/application-sqsadmin.yml (hunk 7 above)
Comment (rschallapalli) [reply src/test/resources/application-sqsadmin.yml]: addressed

**Review by jhevans — APPROVED** [review]
